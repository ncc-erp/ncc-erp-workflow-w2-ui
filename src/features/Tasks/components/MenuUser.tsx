import { Box } from '@chakra-ui/react';
import { FilterAll } from 'common/constants';
import { useMemo } from 'react';
import Select, { SingleValue } from 'react-select';
import { useUserIdentity } from 'api/apiHooks/userIdentityHooks';
import { FilterUserParams } from 'models/userIdentity';
import { UserSortField } from 'common/enums';
import { FilterTasks } from 'models/task';

const initialFilterUser: FilterUserParams = {
  filter: '',
  maxResultCount: 1000, //FIXME: Currently getting 1000 users
  skipCount: 0,
  sorting: [UserSortField.userName, 'asc'].join(' '),
};

interface Props {
  filter: FilterTasks;
  onChange: (
    e: SingleValue<{
      value: string | undefined;
      label: string;
    }>
  ) => void;
}

export const MenuUser = ({ filter, onChange }: Props) => {
  const { data: listUser } = useUserIdentity(initialFilterUser);

  const userOptions = useMemo(() => {
    const defaultOptions = {
      value: '',
      label: FilterAll.USER,
    };

    const options =
      listUser?.items?.map(({ email }) => ({
        value: email,
        label: email,
      })) ?? [];

    return [defaultOptions, ...options];
  }, [listUser]);

  return (
    <Box w={'300px'}>
      <Select
        styles={{
          control: (baseStyles) => ({
            ...baseStyles,
            borderRadius: '0.375rem',
            borderColor: 'inherit',
            paddingBottom: 1,
            fontSize: 14,
          }),
        }}
        value={{
          value: filter.email,
          label: filter.email || FilterAll.USER,
        }}
        options={userOptions}
        onChange={onChange}
      />
    </Box>
  );
};
