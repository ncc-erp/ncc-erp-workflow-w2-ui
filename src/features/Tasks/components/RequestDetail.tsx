import { useOffices } from 'api/apiHooks/requestHooks';
import { TextGroup } from 'common/components/TextGroup/TextGroup';
import { requestTemplateWorkflow } from 'common/constants';
import { IRequest } from 'models/task';
import { extractContent } from 'utils/extractContent';

interface IRequestProps {
  type: string | undefined;
  inputRequestDetail: IRequest | undefined;
}

export const RequestDetail = ({ type, inputRequestDetail }: IRequestProps) => {
  const { data: offices } = useOffices();
  const RequestComponent = {
    [requestTemplateWorkflow.WFH_REQUEST]: (
      <>
        <TextGroup
          label="Current office"
          content={
            offices?.find(
              (office) => office.code === inputRequestDetail?.CurrentOffice
            )?.displayName
          }
        />
        <TextGroup label="Project" content={inputRequestDetail?.Project} />
        <TextGroup
          label="Reason"
          content={extractContent(inputRequestDetail?.Reason)}
        />
        <TextGroup
          label="Dates"
          dates={
            inputRequestDetail?.Dates
              ? inputRequestDetail?.Dates.split(',')
              : []
          }
        />
      </>
    ),
    [requestTemplateWorkflow.OFFICE_EQUIPMENT]: (
      <>
        <TextGroup
          label="Current office"
          content={
            offices?.find(
              (office) => office.code === inputRequestDetail?.CurrentOffice
            )?.displayName
          }
        />
        <TextGroup label="Equipment" content={inputRequestDetail?.Equipment} />
        <TextGroup
          label="Reason"
          content={extractContent(inputRequestDetail?.Reason)}
        />
      </>
    ),
    [requestTemplateWorkflow.CHANGE_OFFICE]: (
      <>
        <TextGroup
          label="Current office"
          content={
            offices?.find(
              (office) => office.code === inputRequestDetail?.CurrentOffice
            )?.displayName
          }
        />
        <TextGroup
          label="Destination Office"
          content={inputRequestDetail?.DestinationOffice}
        />
        <TextGroup
          label="Content"
          content={extractContent(inputRequestDetail?.Content)}
        />
        <TextGroup label="Start date" content={inputRequestDetail?.StartDate} />
        <TextGroup label="End date" content={inputRequestDetail?.EndDate} />
      </>
    ),
    [requestTemplateWorkflow.DEVICE_REQUEST]: (
      <>
        <TextGroup
          label="Current office"
          content={
            offices?.find(
              (office) => office.code === inputRequestDetail?.CurrentOffice
            )?.displayName
          }
        />
        <TextGroup label="Project" content={inputRequestDetail?.Project} />
        <TextGroup label="Device" content={inputRequestDetail?.Device} />
        <TextGroup
          label="Reason"
          content={extractContent(inputRequestDetail?.Reason)}
        />
      </>
    ),
  };

  return type && RequestComponent[type];
};
