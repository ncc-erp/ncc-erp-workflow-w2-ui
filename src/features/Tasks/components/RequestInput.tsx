import { TextGroup } from 'common/components/TextGroup/TextGroup';
import { convertToCase, extractContent, parseDateStrings } from 'utils';

interface IRequestProps {
  inputRequestDetail: Record<string, Record<string, string> | string>;
}

export const RequestInput = ({ inputRequestDetail }: IRequestProps) => {
  const TextGroupRequest = Object.keys(inputRequestDetail).map((key) => {
    const value = inputRequestDetail[key];

    if (typeof value === 'object' && value !== null) {
      const { displayName } = value as { displayName?: string };
      if (displayName) {
        return (
          <TextGroup
            key={key}
            label={convertToCase(key)}
            content={displayName}
          />
        );
      }
    } else {
      const dates = parseDateStrings(value);
      return (
        <TextGroup
          key={key}
          label={convertToCase(key)}
          content={extractContent(value)}
          dates={dates}
        />
      );
    }
  });

  return <>{TextGroupRequest}</>;
};
