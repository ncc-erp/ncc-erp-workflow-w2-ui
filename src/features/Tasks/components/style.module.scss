.container {
    display: flex;
    flex-direction: column;
    margin-top: 10px;
}

.wrapper {
    display: flex;
    width: 100%;
    gap: 0; /* Không có khoảng cách theo trục Y */
    flex-direction: row;
    flex-wrap: wrap;
    column-gap: 2rem; /* Khoảng cách theo trục X là 2rem */
  }

.wrapper > div {
    width: 100%;
    margin-bottom: 10px !important;
}

@media (min-width: 768px) {
    .wrapper  > div {
        width: 46%;
    }
}

.modalBody {
    max-height: 70vh;
    overflow-y: scroll;
}

.containerSpinner {
    display: flex;
    justify-content: center;
    align-items: center;
    margin-top: 10px;
    min-width: 30vw;
    min-height: 50vh;
}

.actions {
    display: flex;
    align-items: center;
    flex-direction: row;
    width: 100%;
    flex-wrap: wrap;
    gap: 5px;
    margin-top: 15px;

    .spinner {
        min-width: 24px;
        margin-right: 5px;
    }
}