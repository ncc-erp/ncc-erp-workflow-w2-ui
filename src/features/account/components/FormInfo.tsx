import {
  FormControl,
  FormLabel,
  Divider,
  Heading,
  Button,
  Input,
  Flex,
  Box,
} from "@chakra-ui/react";
import { yupResolver } from "@hookform/resolvers/yup";
import { useForm } from "react-hook-form";
import { personalInfoSchema } from "utils/yup";

interface IFormPersonal {
  email: string;
  name?: string;
  phoneNumber?: string;
  surname?: string;
  userName: string;
}

const FormInfo = () => {
  const methods = useForm<IFormPersonal>({
    resolver: yupResolver(personalInfoSchema),
    // defaultValues: {
    //   samples: [{ volume: "" }],
    // },
  });
  const {
    handleSubmit,
    register,
    formState: { errors, isSubmitting },
  } = methods;

  function onSubmit(values: IFormPersonal) {
    console.log(values);
  }
  return (
    <>
      <Heading>Personal info</Heading>
      <Divider mb={5} />
      <form onSubmit={handleSubmit(onSubmit)}>
        <FormControl>
          <Flex flexDirection={"column"} gap={5}>
            <Box>
              <FormLabel htmlFor="userName">Username *</FormLabel>
              <Input
                isInvalid={errors?.userName as boolean | undefined}
                id="userName"
                placeholder="Username"
                {...register("userName")}
              />
              <Box fontSize={14} color={"red.500"} position={"absolute"}>
                {errors?.userName && (errors?.userName?.message as string)}
              </Box>
            </Box>
            <Flex gap={5}>
              <Box w={"50%"}>
                <FormLabel htmlFor="name">Name</FormLabel>
                <Input id="name" placeholder="Name" />
              </Box>
              <Box w={"50%"}>
                <FormLabel htmlFor="surname">Surname</FormLabel>
                <Input id="surname" placeholder="Surname" />
              </Box>
            </Flex>
            <Box>
              <FormLabel htmlFor="email">Email *</FormLabel>
              <Input
                isInvalid={errors?.email as boolean | undefined}
                id="email"
                placeholder="Email"
                {...register("email")}
              />
              <Box fontSize={14} color={"red.500"} position={"absolute"}>
                {errors?.email && (errors?.email?.message as string)}
              </Box>
            </Box>
            <Box>
              <FormLabel htmlFor="phone">Phone number</FormLabel>
              <Input id="phone" placeholder="Phone number" />
            </Box>
          </Flex>
        </FormControl>
        <Button
          mt={7}
          colorScheme="blue"
          isLoading={isSubmitting}
          type="submit"
        >
          Submit
        </Button>
      </form>
    </>
  );
};

export default FormInfo;
