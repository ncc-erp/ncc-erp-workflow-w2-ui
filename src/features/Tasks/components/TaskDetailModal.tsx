import {
  Button,
  Divider,
  HStack,
  Heading,
  Image,
  List,
  ListItem,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  Spinner,
  Text,
} from '@chakra-ui/react';
import {
  useActionTask,
  useApproveTask,
  useGetAllTask,
  useGetTaskDetail,
  useRejectTask,
} from 'api/apiHooks/taskHooks';
import Logo from 'assets/images/ncc_logo.png';
import { toast } from 'common/components/StandaloneToast';
import { TextGroup } from 'common/components/TextGroup/TextGroup';
import { WorkflowModal } from 'common/components/WorkflowModal';
import {
  DEFAULT_TASK_PER_PAGE,
  OtherActionSignalStatus,
  TaskStatus,
  UPDATED_BY_W2,
} from 'common/constants';
import { useCallback, useEffect, useMemo, useState } from 'react';
import {
  convertToCase,
  formatDate,
  getStatusByIndex,
  isObjectEmpty,
  subtractTime,
} from 'utils';
import { RequestInput } from './RequestInput';
import { IOtherTasks } from './TasksBoard';
import styles from './style.module.scss';
import { RequestStatus } from 'common/enums';
import { useUserList } from 'api/apiHooks/requestHooks';
import { removeDiacritics } from 'utils/removeDiacritics';
import { BiPencil } from 'react-icons/bi';
import ModalBoard from 'common/components/Boards/ModalBoard';
import { useClearCacheTask } from 'common/components/Boards/useClearCacheTask';
import { FilterTasks } from 'models/task';
import { useCurrentUser } from 'hooks/useCurrentUser';

const initialFilter: FilterTasks = {
  skipCount: 0,
  maxResultCount: DEFAULT_TASK_PER_PAGE,
  workflowDefinitionId: '',
  status: -1,
  dates: subtractTime('months', 1),
  emailRequest: '',
  emailAssign: '',
};

interface IDetailModalProps {
  isOpen: boolean;
  onClose: () => void;
  taskId: string;
  otherTasks?: IOtherTasks;
}

interface IDynamicDataProps {
  [key: string]: string;
}

interface IDynamicReviewProps {
  title: string;
  items: IDynamicDataProps[];
}

export const TaskDetailModal = ({
  isOpen,
  onClose,
  taskId,
  otherTasks,
}: IDetailModalProps) => {
  const { data: users } = useUserList();
  const actionTaskMutation = useActionTask();
  const {
    data,
    refetch,
    isLoading: hasGetTaskLoading,
  } = useGetTaskDetail(taskId);

  const [isLoading, setIsLoading] = useState(false);
  const approveTaskMutation = useApproveTask();
  const rejectTaskMutation = useRejectTask();
  const { clear } = useClearCacheTask();
  const user = useCurrentUser();
  const [isLoadingBtnApprove, setIsLoadingBtnApprove] = useState(false);
  const [isLoadingBtnReject, setIsLoadingBtnReject] = useState(false);
  const [loadStatus] = useState<boolean>(false);
  const [isRejected, setIsRejected] = useState<boolean>(false);

  const [requestWorkflow, setRequestWorkflow] = useState<string>('');
  const [isOpenWorkflow, setOpenWorkflow] = useState(false);

  const onActionViewWorkflow = (workflowId: string) => () => {
    setRequestWorkflow(workflowId);
    setOpenWorkflow(true);
  };

  const [filter] = useState<FilterTasks>({
    ...initialFilter,
    emailAssign: user?.email,
  });

  const [isModalOpen, setIsModalOpen] = useState(false);
  const [modalType, setModalType] = useState<'approve' | 'reject' | ''>('');
  const [reason, setReason] = useState('');
  const [dynamicForm, setDynamicForm] = useState({
    hasDynamicForm: false,
    dynamicForm: '',
  });
  const { refetch: refetchPending } = useGetAllTask(
    { ...filter },
    TaskStatus.Pending
  );

  const { refetch: refetchApproved } = useGetAllTask(
    { ...filter },
    TaskStatus.Approved
  );

  const { refetch: refetchRejected } = useGetAllTask(
    { ...filter },
    TaskStatus.Rejected
  );

  useEffect(() => {
    if (data) {
      setRequestWorkflow(data.tasks.workflowInstanceId || '');
    }
  }, [data]);

  const handleApproveClick = () => {
    setModalType('approve');
    setIsModalOpen(true);
    setIsRejected(false);
  };
  const handleRejectClick = () => {
    setModalType('reject');
    setIsModalOpen(true);
    setIsRejected(true);
  };

  const rejectTask = async (id: string | null) => {
    if (!reason) return;
    setIsLoadingBtnReject(true);
    await rejectTaskMutation
      .mutateAsync({
        id: id as string,
        reason,
      })
      .then(() => {
        toast({ title: 'Rejected Task Successfully!', status: 'success' });
      })
      .catch((error) => {
        console.error(error.response.data.error.message);
      });
    clear();
    refetch();
    refetchRejected();
    refetchPending();
    setIsLoadingBtnApprove(false);
    onClose();
  };

  const approveTask = async (
    id: string | null,
    approvedData?: string | null
  ) => {
    setIsLoadingBtnApprove(true);
    await approveTaskMutation
      .mutateAsync({
        id: id as string,
        dynamicActionData: approvedData,
      })
      .then(() => {
        toast({ title: 'Approved Task Successfully!', status: 'success' });
      })
      .catch((error) => {
        console.error(error.response.data.error.message);
      });
    clear();
    refetch();
    refetchApproved();
    refetchPending();
    setIsLoadingBtnApprove(false);
    onClose();
  };

  const handleModalConfirm = async (data?: string) => {
    setIsModalOpen(false);
    if (modalType === 'approve') {
      await approveTask(taskId, data);
    } else if (modalType === 'reject') {
      await rejectTask(taskId);
    }
  };

  const { tasks, inputRequestUser, inputRequestDetail, emailTo } =
    useMemo(() => {
      const { input, tasks, emailTo } = data || {};
      const { RequestUser, Request } = input || {};

      return {
        tasks,
        emailTo,
        inputRequestUser: RequestUser,
        inputRequestDetail: Request,
      };
    }, [data]);

  const hasTaskAction: boolean = useMemo(() => {
    if (!data?.tasks || typeof data.tasks.status !== 'number') {
      return false;
    }

    return (
      data.tasks.status === TaskStatus.Pending &&
      !!data.otherActionSignals?.length
    );
  }, [data]);

  const hasInputRequestData: boolean = useMemo(() => {
    return !isObjectEmpty(inputRequestDetail);
  }, [inputRequestDetail]);

  const onActionClick = async (id: string, action: string) => {
    try {
      setIsLoading(true);
      await actionTaskMutation.mutateAsync({ id, action });
      toast({ title: 'Send action successfully!', status: 'success' });
      refetch();
    } catch (error) {
      console.error(error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    if (isOpen) {
      setIsLoadingBtnApprove(false);
      setIsLoadingBtnReject(false);
      if (isRejected) {
        setDynamicForm({
          hasDynamicForm: false,
          dynamicForm: '',
        });
      } else {
        const { dynamicActionData } = tasks || {};
        setDynamicForm({
          hasDynamicForm: !!dynamicActionData,
          dynamicForm: dynamicActionData || '',
        });
      }
    }
  }, [isOpen, isRejected, tasks]);

  const convertToDynamicArray = (payload: string | null | undefined) => {
    if (!payload) return [];

    try {
      const data = JSON.parse(payload) as IDynamicDataProps[];
      return data.map((element) => ({
        data: (element.data || '').split('\n'),
        name: element.name || 'No Name',
      })) as unknown as IDynamicDataProps[];
    } catch (error) {
      return [];
    }
  };

  const getUserReject = useMemo(() => {
    if (
      (getStatusByIndex(tasks?.status).status as string) !==
      RequestStatus.Rejected
    ) {
      return null;
    }

    const otherTasksSorted = otherTasks?.items?.sort(
      (a, b) =>
        new Date(b?.creationTime).getTime() -
        new Date(a?.creationTime).getTime()
    );

    const userReject = otherTasksSorted?.find(
      (task) => task.updatedBy != null && task.updatedBy != UPDATED_BY_W2
    )?.updatedBy;

    return users?.find((user) => user.email == userReject)?.name;
  }, [tasks?.status, otherTasks, users]);

  const mappingReviewToList = (data: IDynamicDataProps[]) => {
    return data.map((element, ind) => {
      if (!Array.isArray(element.data)) return null;

      const filteredData = element.data.filter((item) => item.trim() !== '');

      if (filteredData.length === 0) return null;

      return (
        <List key={element.name + ind} mt={1} spacing={1}>
          <Text fontSize={14} fontWeight={600} fontStyle="italic">
            {convertToCase(element.name)}:
          </Text>
          {filteredData.map((x) => (
            <ListItem key={x}>{x}</ListItem>
          ))}
        </List>
      );
    });
  };

  const renderDynamicDataContent = useCallback(() => {
    if (!otherTasks || otherTasks.items.length <= 0) return null;

    const filterOtherTask: IDynamicReviewProps[] = otherTasks.items.map((x) => {
      return {
        title: `${x.description || 'No name'} (${x.updatedBy
          ?.split('@')
          .shift()})`,
        items: convertToDynamicArray(x.dynamicActionData),
      };
    });

    const tasksWithData = filterOtherTask.filter((task) =>
      task.items.some(
        (item) =>
          Array.isArray(item.data) &&
          item.data.some((data) => data.trim() !== '')
      )
    );

    return tasksWithData.map((x, ind) => {
      return (
        <div key={ind}>
          <Text
            display="flex"
            alignItems="center"
            gap={1}
            fontSize={15}
            mt={2}
            fontWeight={600}
          >
            {x.title} <BiPencil fontSize={15} />
          </Text>
          {mappingReviewToList(x.items)}
        </div>
      );
    });
  }, [otherTasks]);

  if (hasGetTaskLoading) {
    return (
      <Modal isOpen={isOpen} onClose={onClose}>
        <ModalOverlay />
        <ModalContent p="10px" maxW="700px">
          <ModalBody>
            <div className={styles.containerSpinner}>
              <Spinner color="red.500" size="xl" />
            </div>
          </ModalBody>
        </ModalContent>
      </Modal>
    );
  }

  return (
    <>
      <Modal isOpen={isOpen} onClose={onClose} blockScrollOnMount={false}>
        <ModalOverlay />
        <ModalContent p="10px" maxW="700px">
          <ModalHeader>
            <HStack>
              <Image h="45px" src={Logo} />
              <Heading ml={1} w="550px">
                <Text color="primary" fontSize={18}>
                  {tasks?.name}
                </Text>
                <Text fontSize={16} fontWeight={400} mt={1.5}>
                  {tasks?.description}
                </Text>
              </Heading>
            </HStack>
            <div className={styles.listActionContainer}>
              <Button
                className={styles.btnWFDetail}
                onClick={onActionViewWorkflow(
                  data?.tasks.workflowInstanceId as string
                )}
                mt={2}
              >
                View Workflow Detail
              </Button>
              <div className={styles.actions}>
                <div className={styles.spinner}>
                  {isLoading && <Spinner color="red.500" />}
                </div>
                {hasTaskAction &&
                  data?.otherActionSignals
                    ?.sort((a, b) => {
                      return a.otherActionSignal.localeCompare(
                        b.otherActionSignal
                      );
                    })
                    .map((x, ind) => {
                      return (
                        <Button
                          key={ind}
                          isDisabled={
                            isLoading ||
                            x.status !== OtherActionSignalStatus.PENDING
                          }
                          onClick={() =>
                            onActionClick(data?.tasks.id, x.otherActionSignal)
                          }
                        >
                          {x.otherActionSignal}
                        </Button>
                      );
                    })}
              </div>
              <div className={styles.listBtnActionTask}>
                {tasks?.status === TaskStatus.Approved ||
                tasks?.status === TaskStatus.Rejected ? null : (
                  <>
                    <Button
                      colorScheme="green"
                      isLoading={isLoadingBtnApprove}
                      onClick={handleApproveClick}
                      mt={2}
                      className={styles.btnApproveTask}
                    >
                      Approve
                    </Button>
                    <Button
                      colorScheme="red"
                      isLoading={isLoadingBtnReject}
                      onClick={handleRejectClick}
                      mt={2}
                      className={styles.btnRejectTask}
                    >
                      Reject
                    </Button>
                  </>
                )}
              </div>
            </div>
          </ModalHeader>
          <ModalCloseButton mt="15px" mr="10px" />
          <ModalBody className={styles.modalBody} pr={2}>
            <Divider mt={2} mb={3} />
            <div className={styles.container}>
              <Text
                mb="10px"
                fontWeight={600}
                fontStyle="italic"
                color="primary"
              >
                Request input
              </Text>

              <div className={styles.wrapper}>
                {hasInputRequestData && inputRequestDetail && (
                  <RequestInput inputRequestDetail={inputRequestDetail} />
                )}
              </div>

              <Divider mt={2} mb={3} />
              <Text
                mb="10px"
                fontWeight={600}
                fontStyle="italic"
                color="primary"
              >
                Request user
              </Text>

              <div className={styles.wrapper}>
                <TextGroup label="Name" content={inputRequestUser?.name} />
                <TextGroup label="Email" content={inputRequestUser?.email} />
                <TextGroup
                  label="Branch name"
                  content={inputRequestUser?.branchName}
                />
              </div>

              <Divider mt={2} mb={3} />
              <Text
                mb="15px"
                fontWeight={600}
                fontStyle="italic"
                color="primary"
              >
                Detail
              </Text>
              <div className={styles.wrapper}>
                <TextGroup label="Request template" content={tasks?.name} />
                <TextGroup
                  label="State"
                  content={getStatusByIndex(tasks?.status).status}
                  color={getStatusByIndex(tasks?.status).color}
                />
                {tasks?.reason && (
                  <TextGroup label="Reason" content={tasks.reason} />
                )}

                <TextGroup
                  label="Email assignment"
                  content={emailTo?.join(', ')}
                />
                <TextGroup
                  label="Creation time"
                  content={
                    tasks?.creationTime
                      ? formatDate(new Date(tasks?.creationTime))
                      : ''
                  }
                />
                {getUserReject && (
                  <TextGroup
                    label="Rejected by"
                    content={removeDiacritics(getUserReject)}
                  />
                )}
                {renderDynamicDataContent()}
              </div>
            </div>
          </ModalBody>
        </ModalContent>
      </Modal>
      <ModalBoard
        isOpen={isModalOpen}
        onClose={() => setIsModalOpen(false)}
        onConfirm={handleModalConfirm}
        showReason={isRejected}
        showDynamicForm={dynamicForm.hasDynamicForm}
        dynamicForm={dynamicForm.dynamicForm}
        setReason={setReason}
        shortTitle={tasks?.title}
        isLoading={loadStatus}
        name={tasks?.name}
        requestUser={data?.input?.RequestUser?.name}
        isDisabled={isRejected && !reason}
      />
      {requestWorkflow && (
        <WorkflowModal
          isOpen={isOpenWorkflow}
          onClose={() => setOpenWorkflow(false)}
          workflow={`CompOnly?id=${requestWorkflow}`}
        />
      )}
    </>
  );
};
