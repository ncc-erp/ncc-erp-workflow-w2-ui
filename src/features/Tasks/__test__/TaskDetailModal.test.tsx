import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { TaskDetailModal } from '../components/TaskDetailModal';

jest.mock('../../../api/apiHooks/index', () => ({
  useAxios: jest.fn(),
}));

jest.mock('../../../api/axiosInstant', () => ({
  VITE_API_BASE_URL: '/api',
}));

jest.mock('common/components/WorkflowModal', () => ({
  VITE_PROXY_SERVER_URL: 'http://localhost:4433',
}));

jest.mock('hooks/useIsAdmin', () => ({
  useIsAdmin: jest.fn().mockReturnValue(true),
}));

jest.mock('api/apiHooks/requestHooks', () => ({
  useOffices: jest.fn().mockReturnValue({
    data: [
      {
        displayName: 'Đà Nẵng',
        code: 'ĐN',
        headOfOfficeEmail: 'thien.dang@ncc.asia',
      },
    ],
  }),
}));

jest.mock('api/apiHooks/taskHooks', () => ({
  useActionTask: jest.fn(() => ({
    mutateAsync: jest.fn(),
  })),
  useGetTaskDetail: jest.fn().mockReturnValue({
    refetch: jest.fn(),
    isLoading: false,
    data: {
      tasks: {
        id: '3a0e2af0-769c-6fa4-ccab-90ca0afe6861',
        workflowInstanceId: '3a0e2af0-73ad-ff31-9f9c-6f5d347d48ec',
        workflowDefinitionId: '3a0e256f-a848-065c-fab4-56e8c133918b',
        email: null,
        status: 0,
        name: 'Probationary Request',
        description: 'PM Reviews',
        dynamicActionData:
          '[{"name":"StrengthPoints","type":"RichText","isRequired":true,"data":""},{"name":"WeaknessPoints","type":"RichText","isRequired":true,"data":""}]',
        reason: null,
        creationTime: '2023-10-10T08:45:55.228848Z',
        otherActionSignals: [],
        emailTo: null,
        author: '3a0d3d06-ffbb-0487-b6e4-9b3488ecec70',
        authorName: null,
      },
      emailTo: ['long.vodinhhoang@ncc.asia', 'trang.danghuyen@ncc.asia'],
      otherActionSignals: null,
      input: {
        Request: {
          content: 'sdasda',
          currentOffice: {
            displayName: 'Đà Nẵng',
            code: 'ĐN',
            headOfOfficeEmail: 'thien.dang@ncc.asia',
          },
          startDate: '04/10/2023',
          endDate: '01/02/2024',
        },
        RequestUser: {
          email: 'long.vodinhhoang@ncc.asia',
          name: 'Long Vo Dinh Hoang',
          project: null,
          pm: 'thien.dang@ncc.asia',
          headOfOfficeEmail: 'thien.dang@ncc.asia',
          projectCode: 'ncc-w2',
          branchName: 'Đà Nẵng',
          branchCode: 'ĐN',
          id: '3a0d3d06-ffbb-0487-b6e4-9b3488ecec70',
        },
      },
    },
  }),
}));

test('Task Detail Modal', () => {
  const queryClient: QueryClient = new QueryClient();
  const { baseElement } = render(
    <QueryClientProvider client={queryClient}>
      <TaskDetailModal
        isOpen={true}
        onClose={() => jest.fn()}
        taskId="abcxyz"
      />
    </QueryClientProvider>
  );
  expect(baseElement).toMatchSnapshot();
});
