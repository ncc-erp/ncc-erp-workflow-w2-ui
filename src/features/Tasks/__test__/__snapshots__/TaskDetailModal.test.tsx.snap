// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Detail Modal 1`] = `
<body>
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  />
  <div
    class="chakra-portal"
  >
    <div
      aria-hidden="true"
      class="chakra-modal__overlay css-13d1tkw"
      data-aria-hidden="true"
      style="opacity: 0;"
    />
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="chakra-modal__content-container css-1oxhx2p"
        tabindex="-1"
      >
        <section
          aria-describedby="chakra-modal--body-:r0:"
          aria-labelledby="chakra-modal--header-:r0:"
          aria-modal="true"
          class="chakra-modal__content css-ctar5y"
          id="chakra-modal-:r0:"
          role="dialog"
          style="opacity: 0; transform: scale(0.95) translateZ(0);"
          tabindex="-1"
        >
          <header
            class="chakra-modal__header css-72fd9l"
            id="chakra-modal--header-:r0:"
          >
            <div
              class="chakra-stack css-1igwmid"
            >
              <img
                class="chakra-image css-kow2w3"
                src=""
              />
              <h2
                class="chakra-heading css-sigo6f"
              >
                <p
                  class="chakra-text css-fb81nc"
                >
                  Probationary Request
                </p>
                <p
                  class="chakra-text css-issjhg"
                >
                  PM Reviews
                </p>
              </h2>
            </div>
            <div
              class="listActionContainer"
            >
              <button
                class="chakra-button btnWFDetail css-12eh3jv"
                type="button"
              >
                View Workflow Detail
              </button>
              <div
                class="actions"
              >
                <div
                  class="spinner"
                />
              </div>
              <div
                class="listBtnActionTask"
              >
                <button
                  class="chakra-button btnApproveTask css-12eh3jv"
                  type="button"
                >
                  Approve
                </button>
                <button
                  class="chakra-button btnRejectTask css-12eh3jv"
                  type="button"
                >
                  Reject
                </button>
              </div>
            </div>
          </header>
          <button
            aria-label="Close"
            class="chakra-modal__close-btn css-1m7215s"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-onkibi"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="chakra-modal__body modalBody css-1qou8as"
            id="chakra-modal--body-:r0:"
          >
            <hr
              aria-orientation="horizontal"
              class="chakra-divider css-1urc9z9"
            />
            <div
              class="container"
            >
              <p
                class="chakra-text css-10g9qth"
              >
                Request input
              </p>
              <div
                class="wrapper"
              >
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Content
                  </label>
                  <p
                    class="content "
                  >
                    <div>
                      Mocked MarkdownEditor
                    </div>
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Current Office
                  </label>
                  <p
                    class="content "
                  >
                    Đà Nẵng
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Start Date
                  </label>
                  <div
                    class="dates"
                  >
                    <p
                      class="date"
                    >
                      04/10/2023
                    </p>
                  </div>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    End Date
                  </label>
                  <div
                    class="dates"
                  >
                    <p
                      class="date"
                    >
                      01/02/2024
                    </p>
                  </div>
                </div>
              </div>
              <hr
                aria-orientation="horizontal"
                class="chakra-divider css-1urc9z9"
              />
              <p
                class="chakra-text css-10g9qth"
              >
                Request user
              </p>
              <div
                class="wrapper"
              >
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Name
                  </label>
                  <p
                    class="content "
                  >
                    Long Vo Dinh Hoang
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Email
                  </label>
                  <p
                    class="content "
                  >
                    long.vodinhhoang@ncc.asia
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Branch name
                  </label>
                  <p
                    class="content "
                  >
                    Đà Nẵng
                  </p>
                </div>
              </div>
              <hr
                aria-orientation="horizontal"
                class="chakra-divider css-1urc9z9"
              />
              <p
                class="chakra-text css-yb4mb5"
              >
                Detail
              </p>
              <div
                class="wrapper"
              >
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Request template
                  </label>
                  <p
                    class="content "
                  >
                    Probationary Request
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    State
                  </label>
                  <p
                    class="content content-blue"
                  >
                    Pending
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Email assignment
                  </label>
                  <p
                    class="content "
                  >
                    long.vodinhhoang@ncc.asia, trang.danghuyen@ncc.asia
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Creation time
                  </label>
                  <p
                    class="content "
                  >
                    10-10-2023 15:45
                  </p>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</body>
`;
